contractcourt/channel_arbitrator.go

```diff
@@ -3159,3 +3107,251 @@ func (c *ChannelArbitrator) checkLegacyBreach() (ArbitratorState, error) {
+	remotePendingReq := fn.None[sweepRequest]()
+
+	// First we check on the remote pending commitment and optionally
+	// create an anchor sweeping request.
+	htlcs, ok := c.activeHTLCs[RemotePendingHtlcSet]
+	if ok && anchors.RemotePending != nil {
+		req, err := c.createSweepRequest(
+			anchors.RemotePending, htlcs, "remote pending",
+			heightHint,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// Save the request.
+		requests = append(requests, req)
+
+		// Set the optional variable.
+		remotePendingReq = fn.Some(req)
+	}
+
+	// Check the local commitment and optionally create an anchor sweeping
+	// request. The params used in this request will be influenced by the
+	// anchor sweeping request made from the pending remote commitment.
+	htlcs, ok = c.activeHTLCs[LocalHtlcSet]
+	if ok && anchors.Local != nil {
+		req, err := c.createSweepRequest(
+			anchors.Local, htlcs, "local", heightHint,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		// If there's an anchor sweeping request from the pending
+		// remote commitment, we will compare its budget against the
+		// budget used here and choose the params that has a larger
+		// budget. The deadline when choosing the remote pending budget
+		// instead of the local one will always be earlier or equal to
+		// the local deadline because outgoing HTLCs are resolved on
+		// the local commitment first before they are removed from the
+		// remote one.
+		remotePendingReq.WhenSome(func(s sweepRequest) {
+			if s.params.Budget <= req.params.Budget {
+				return
+			}
+
+			log.Infof("ChannelArbitrator(%v): replaced local "+
+				"anchor(%v) sweep params with pending remote "+
+				"anchor sweep params, \nold:[%v], \nnew:[%v]",
+				c.cfg.ChanPoint, anchors.Local.CommitAnchor,
+				req.params, s.params)
+
+			req.params = s.params
+		})
+
+		// Save the request.
+		requests = append(requests, req)
+	}
+
+	// Check the remote commitment and create an anchor sweeping request if
+	// needed.
+	htlcs, ok = c.activeHTLCs[RemoteHtlcSet]
+	if ok && anchors.Remote != nil {
+		req, err := c.createSweepRequest(
+			anchors.Remote, htlcs, "remote", heightHint,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		requests = append(requests, req)
+	}
+
+	return requests, nil
+}
+
+// cancelIncomingHTLCs cancels back the incoming htlcs for the corresponding
+// outgoing htlc. An example could be dust htlcs which cannot be enforced
+// onchain therefore we can immediately cancel back the corresponding incoming
+// htlc even if the commitment tx is still not confirmed and there is a risk of
+// another state confirming e.g. the remote pending commitment. However we
+// might in the future cancel back non-dust HTLCs that's why this function is
+// kept generic and not only for dust HTLCs.
+func (c *ChannelArbitrator) cancelIncomingHTLCs(
+	htlcs fn.Set[uint64]) error {
+
+	log.Debugf("ChannelArbitrator(%v): cancelling back %v incoming "+
+		"HTLC(s), includes also dust HTLCs", c.cfg.ChanPoint,
```

Comment:
a bit confusing since we are not doing non-dust canceling atm, think we should just mention we are canceling back dust incoming HTLCs

Reply:
but we are also cancel non-dust dangling here.